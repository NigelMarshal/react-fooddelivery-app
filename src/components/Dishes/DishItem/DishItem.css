.dish__item--selected::before {
  border-style: solid;
  border-left: solid 7px;
  border-color: #1258ff;
  margin-right: 0.5em;
  height: 130px;
  content: "";
  position: relative;
  margin-left: -14px;
}
