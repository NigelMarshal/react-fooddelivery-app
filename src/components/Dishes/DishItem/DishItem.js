import './DishItem.css';
import React, { useState } from 'react'
import StockAlert from '../../StockAlert/StockAlert'
import useLocalStorage from '../../../hooks/useLocalStorage'

const DishItem = ({ menuItems }) => {

    const [selected, setSelected] = useState();
    const [selectedCount, setSelectedCount] = useLocalStorage(menuItems.name, 0);
    const [isOpen, setIsOpen] = useState(false);

    const handleMenuItem = () => {
        //On click toggle selector
        setSelected(true);
        //Per click, this goes down until it runs out for that item
        if (menuItems.dishesInStock) {
            menuItems.dishesInStock -= 1;
            setSelectedCount(selectedCount + 1)
        }
        //Toggle a modal that notifies when you've hit the order/inventory limit
        if (menuItems.dishesInStock === 0) {
            togglePopup();
        }
    }

    const togglePopup = () => {
        setIsOpen(!isOpen);
    }
    return (
        <>
            {/* Check for inventory/cart status */}
            {isOpen && <StockAlert />}
            <div className={selected || selectedCount > 0 ? "py-4 flex dish__item--selected cursor-pointer" : "py-4 flex cursor-pointer"}
                onClick={handleMenuItem}
                data-testid="dishItems"
            >
                <div className="w-9/12">
                    {/* Display the counter for each selected dish */}
                    <p className="text-xl font-medium text-black semi-bold">{(selectedCount !== 0) && `${selectedCount}x`} {menuItems.name}</p>
                    <p className="text-gray pb-5">{menuItems.description}</p>
                    {/* Calculating price and conditionally rendering the discount. 
                    Prices are rounded off to the nearest integer */}
                    {
                        (menuItems.discount_rate !== 0) &&
                        <p className="inline-block pr-3 text-black semi-bold"> AED {Math.round((menuItems.price / 100) - (menuItems.discount_rate * (menuItems.price / 100)))} </p>

                    }
                    <p className={menuItems.discount_rate === 0 ? "pb-5 inline-block" : "pb-5 inline-block line-through text-gray"}>AED {menuItems.price / 100}</p>

                </div>
                <div className="ml-auto">
                    {
                        (menuItems.photo) && <img className="rounded-xl" src={menuItems.photo} width="91" height="91" alt="" />
                    }
                </div>
            </div>
            <hr className="light-gray" />
        </>
    )
}

export default DishItem
