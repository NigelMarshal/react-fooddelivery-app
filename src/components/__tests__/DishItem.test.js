import React from 'react';
import { render, screen, cleanup } from '@testing-library/react'
import renderer from "react-test-renderer";
import DishItem from '../Dishes/DishItem/DishItem';
import dishes from './mocks/dishes'
import Enzyme, { shallow } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';

Enzyme.configure({ adapter: new Adapter() });

afterEach(() => {
    cleanup();
});

test('Should render dish Items', () => {
    render(<DishItem menuItems={dishes[0]} />);
    const dishItems = screen.getByTestId('dishItems');
    expect(dishItems).toBeInTheDocument();
})

test('Dishes should render its content details', () => {
    const component = renderer.create(
        <DishItem menuItems={dishes[0]} />
    );
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
})

test('Dishes should show the discounted price', () => {
    render(
        <DishItem menuItems={dishes[0]} />
    );
    const price = screen.getByText("AED 50")
    expect(price).toBeInTheDocument()
})








test('Dishes should NOT show discounted price if there is none', () => {
    const wrapper = shallow(<DishItem menuItems={dishes[1]} />)
    expect(wrapper.find('.line-through').exists()).toBe(false);
})
