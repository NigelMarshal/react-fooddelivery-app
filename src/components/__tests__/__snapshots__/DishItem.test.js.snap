// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dishes should render its content details 1`] = `
Array [
  <div
    className="py-4 flex cursor-pointer"
    data-testid="dishItems"
    onClick={[Function]}
  >
    <div
      className="w-9/12"
    >
      <p
        className="text-xl font-medium text-black semi-bold"
      >
         
        Dish1
      </p>
      <p
        className="text-gray pb-5"
      >
        Dish1 description
      </p>
      <p
        className="inline-block pr-3 text-black semi-bold"
      >
         AED 
        25
         
      </p>
      <p
        className="pb-5 inline-block line-through text-gray"
      >
        AED 
        50
      </p>
    </div>
    <div
      className="ml-auto"
    />
  </div>,
  <hr
    className="light-gray"
  />,
]
`;
